jlm/llvm/opt/alias-analyses/MemoryStateEncoder.cpp

```diff
@@ -15,11 +15,140 @@
 namespace jlm::llvm::aa
 {
 
+/**
+ * \brief Helper struct for counting up MemoryNodes, among some set of entities that use them
+ */
+struct MemoryStateTypeCounter final
+{
+  // The number of entities that have been counted
+  uint64_t NumEntities = 0;
+
+  // Count of total memory states, divided by MemoryNode type
+  uint64_t NumAllocas = 0;
+  uint64_t NumMallocs = 0;
+  uint64_t NumDeltas = 0;
+  uint64_t NumImports = 0;
+  uint64_t NumLambdas = 0;
+  uint64_t NumExternal = 0;
+
+  // Among the MemoryNodes counted above, how many were not escaping
+  uint64_t NumNonEscaped = 0;
+
+  // Remember the single entity with the highest number of memory states
+  uint64_t MaxMemoryStateEntity = 0;
+  // Do the same, but only include non-escaped MemoryNodes
+  uint64_t MaxNonEscapedMemoryStateEntity = 0;
+
+  void
+  CountEntity(
+      uint64_t numAllocas,
+      uint64_t numMallocs,
+      uint64_t numDeltas,
+      uint64_t numImports,
+      uint64_t numLambdas,
+      uint64_t numExternal,
+      uint64_t numNonEscaped)
+  {
+    NumEntities++;
+
+    NumAllocas += numAllocas;
+    NumMallocs += numMallocs;
+    NumDeltas += numDeltas;
+    NumImports += numImports;
+    NumLambdas += numLambdas;
+    NumExternal += numExternal;
+
+    const uint64_t totalMemoryStates =
+        numAllocas + numMallocs + numDeltas + numImports + numLambdas + numExternal;
+    if (totalMemoryStates > MaxMemoryStateEntity)
+      MaxMemoryStateEntity = totalMemoryStates;
+
+    NumNonEscaped += numNonEscaped;
+    if (numNonEscaped > MaxNonEscapedMemoryStateEntity)
+      MaxNonEscapedMemoryStateEntity = numNonEscaped;
+  }
+
+  void
+  CountEntity(util::HashSet<const PointsToGraph::MemoryNode *> memoryNodes)
+  {
+    uint64_t numAllocas = 0;
+    uint64_t numMallocs = 0;
+    uint64_t numDeltas = 0;
+    uint64_t numImports = 0;
+    uint64_t numLambdas = 0;
+    uint64_t numExternal = 0;
+
+    uint64_t numNonEscaped = 0;
+
+    for (const auto memoryNode : memoryNodes.Items())
+    {
+      if (!memoryNode->IsModuleEscaping())
+        numNonEscaped++;
+
+      if (dynamic_cast<const PointsToGraph::AllocaNode *>(memoryNode))
+        numAllocas++;
+      else if (dynamic_cast<const PointsToGraph::MallocNode *>(memoryNode))
+        numMallocs++;
+      else if (dynamic_cast<const PointsToGraph::DeltaNode *>(memoryNode))
+        numDeltas++;
+      else if (dynamic_cast<const PointsToGraph::ImportNode *>(memoryNode))
+        numImports++;
+      else if (dynamic_cast<const PointsToGraph::LambdaNode *>(memoryNode))
+        numLambdas++;
+      else if (dynamic_cast<const PointsToGraph::ExternalMemoryNode *>(memoryNode))
+        numExternal++;
+      else
+        JLM_UNREACHABLE("Unknown MemoryNode type");
```

Comment:
I think there is also an is<> method for memory nodes defined.

Reply:
I could not find one, so I added one

Reply:
There is already one: https://github.com/phate/jlm/blob/342848731a725be9648cb99fb5370596a325e0bc/jlm/llvm/opt/alias-analyses/PointsToGraph.hpp#L535C1-L542C4